services/gitdiff/gitdiff.go

```diff
@@ -555,6 +283,182 @@ func DiffInlineWithHighlightCode(fileName, language, code string) DiffInline {
+// These Unicode placeholders are friendly to the diff.
+// Then after diff, the placeholders in diff result will be recovered to the HTML tags.
+// It's guaranteed that the tags in final diff result are paired correctly.
+type HighlightCodeDiff struct {
+	placeholderBegin    rune
+	placeholderMaxCount int
+	placeholderIndex    int
+	placeholderTagMap   map[rune]string
+	tagPlaceholderMap   map[string]rune
+
+	lineWrapperTags []string
+}
+
+func NewHighlightCodeDiff() *HighlightCodeDiff {
+	return &HighlightCodeDiff{
+		placeholderBegin:    rune(0xE000), // Private Use Unicode: U+E000..U+F8FF, BMP(0), 6400
+		placeholderMaxCount: 6400,
+		placeholderTagMap:   map[rune]string{},
+		tagPlaceholderMap:   map[string]rune{},
+	}
+}
+
+// nextPlaceholder returns 0 if no more placeholder can be used
+// the diff is done line by line, usually there are only a few (no more than 10) placeholders in one line
+// so the placeholderMaxCount is impossible to be exhausted in real cases.
+func (hcd *HighlightCodeDiff) nextPlaceholder() rune {
+	for hcd.placeholderIndex < hcd.placeholderMaxCount {
+		r := hcd.placeholderBegin + rune(hcd.placeholderIndex)
+		hcd.placeholderIndex++
+		// only use non-existing (not used by code) rune as placeholders
+		if _, ok := hcd.placeholderTagMap[r]; !ok {
+			return r
+		}
+	}
+	return 0 // no more available placeholder
+}
+
+func (hcd *HighlightCodeDiff) isInPlaceholderRange(r rune) bool {
+	return hcd.placeholderBegin <= r && r < hcd.placeholderBegin+rune(hcd.placeholderMaxCount)
+}
+
+func (hcd *HighlightCodeDiff) collectUsedRunes(code string) {
+	for _, r := range code {
+		if hcd.isInPlaceholderRange(r) {
+			// put the existing rune (used by code) in map, then this rune won't be used a placeholder anymore.
+			hcd.placeholderTagMap[r] = ""
+		}
+	}
+}
+
+func (hcd *HighlightCodeDiff) diffWithHighlight(filename, language, codeA, codeB string) []diffmatchpatch.Diff {
+	hcd.collectUsedRunes(codeA)
+	hcd.collectUsedRunes(codeB)
+
+	highlightCodeA := highlight.Code(filename, language, codeA)
+	highlightCodeB := highlight.Code(filename, language, codeB)
+
+	highlightCodeA = hcd.convertToPlaceholders(highlightCodeA)
+	highlightCodeB = hcd.convertToPlaceholders(highlightCodeB)
+
+	diffs := diffMatchPatch.DiffMain(highlightCodeA, highlightCodeB, true)
+	diffs = diffMatchPatch.DiffCleanupEfficiency(diffs)
+
+	for i := range diffs {
+		hcd.recoverOneDiff(&diffs[i])
+	}
+	return diffs
+}
+
+func (hcd *HighlightCodeDiff) convertToPlaceholders(htmlCode string) string {
+	var tagStack []string
+	res := strings.Builder{}
+
+	firstRunForLineTags := hcd.lineWrapperTags == nil
+
+	// the standard chroma highlight HTML is "<span class="line [hl]"><span class="cl"> ... </span></span>"
+	for {
+		// find the next HTML tag
+		pos1 := strings.IndexByte(htmlCode, '<')
+		pos2 := strings.IndexByte(htmlCode, '>')
+		if pos1 == -1 || pos2 == -1 || pos2 < pos1 {
```

Comment:
We know that `pos2 > pos1` if `pos1 != pos2 != -1` because we're relying on chroma producing valid HTML.

<details><summary>If htmlCode is not valid HTML then all of this code is unsafe and this test is not enough to detect that. </summary>

For example there's nothing to ensure that detected `>` isn't within an attribute e.g.

```
<div class="I am technically invalid >< but actually most HTML parsers would tolerate me">
```

</details>

It might be prudent to add specific comments to these functions explicitly warning that it is assumed that only valid HTML can be passed in and that these sorts of cases are not being looked for.

Reply:
Why should invalid htmlCode be passed into this function? These functions are private functions (lower-cased first latter), only used by internal code.

And they are covered by unit tests, if Chroma changed its behavior and breaks something, this code should also be updated.

What comments do you think is clear? Can you edit on this PR directly?

Reply:
If you don't clearly write what the assumptions are around a piece of code those assumptions will be forgotten and this code will almost certainly be used in an unintended fashion in future. 

One of the problems with adding checks like the referenced check is that sometimes they give the impression that this would be safe for invalid HTML. It would not be safe, as I wrote above, and although we both know this - we may not be the ones reviewing or writing code that uses this in an unsafe manner. 

I'm not saying that this code is incorrect - all I'm saying is add a warning because someone may try to use it in an incorrect way. We should explain why it would be unsafe to use this on generic "html" that was not guaranteed to be valid and why.

Reply:
Do you mean to add a comment like "// It totally depends on Chroma's valid HTML output and its structure, do not use these functions for other purposes." ?